src/DatasetNormalizerValidatorDecorator.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator;
+
+use Closure;
+use JetBrains\PhpStorm\Pure;
+use Psr\Container\ContainerExceptionInterface;
+use Psr\Container\NotFoundExceptionInterface;
+use ReflectionProperty;
+use Traversable;
+use Yiisoft\Validator\DataSet\ArrayDataSet;
+use Yiisoft\Validator\DataSet\MixedDataSet;
+use Yiisoft\Validator\DataSet\ObjectDataSet;
+use Yiisoft\Validator\RulesProvider\AttributesRulesProvider;
+
+use function is_array;
+
+/**
+ * Validator validates {@link DataSetInterface} against rules set for data set attributes.
+ */
+final class DatasetNormalizerValidatorDecorator implements ValidatorInterface
```

Comment:
I still think it doesn't make sense to decouple it from the validator itself since without normalizer validator isn't really usable.